### DISCUSSION TITLE
Request: provide a generic Link to the latest Windows release

### DISCUSSION THREAD

**Discussion by hamlibdk** [git-for-windows/git#3020] [Ideas]
Is it possible to provide a "generic link" to the latest releases (especially Windows releases) so that utilities that use tools such as  cURL or PowerShell Invoke-RestMethod can always point to the same file yet always download the very latest GIT release?

i.e. current Windows release is 2.29.2.3. Versions change frequently. Can a static link - as an example. Git-LATEST-x64.exe -  be provided to this latest release and maintained always to the latest release?

This allows the general programming community to create generic environments where static version coding is eliminated - hence reducing maintenance required on scripts and code. This also allows us to better fit in with current and future licensing requirements

Thanks.

Hamlib SDK Development Team
(Maintainer of the JTSDK for building WSJT-X and forks)

**Comment by PhilipOakley** [comment]
One option is the [git-scm](https://git-scm.com/) page which has an actively updated link https://git-scm.com/download/win 
I believe it the GitHub staff that maintain that page.

**Comment by dscho** [comment]
GitHub Pages does not have redirects, as far as I know. So there isn't any easy way to do that, I don't think. We do update https://gitforwindows.org/latest-version.txt (and `-tag`) though.

**Comment by dscho** [comment]
hrm. I seem to be unable to convert this issue to a discussion. :disappointed:

**Comment by hamlibdk** [comment]
Apologies for lack of response but I have been on limited Internet Resources in Central Australia ... The https://gitforwindows.org/latest-version.txt is highly useful and will be used as the basis for experiments taking place.

**Comment by hamlibdk** [comment]
Further issue....

Yes ... https://gitforwindows.org/latest-version.txt is great... But what about the download release path for this version?

The Path nomenclature is inconsistent in its structure... and cannot be calculated (with ease) as far as I can see easily from just the version data.

My point is the path https://github.com/git-for-windows/git/releases/download/v2.30.0.windows.2/Git-2.30.0.2-64-bit.exe for the latest at the time of publication is NOT intuitive. Inserting the "windows" creates unnecessary complexity especially for those of us trying to write software that is designed to be predominantly "maintenance free" and fulfils all reasonable (and possibly future) legal responsibilities of pulling latest version of software DIRFECTLY from host sites rather than repackaging it !

**Comment by hamlibdk** [comment]
PowerShell Script as deployed..... The only issue is if some nomenclature changes....

```
Set-Location -Path $PSScriptRoot

Write-Host ""
Write-Host "* Downloading Git Installer"

# Get current Git-SCM Version
Write-Host ""
$appName="latest-version.txt"
$appURL="https://gitforwindows.org/latest-version.txt"
Invoke-RestMethod -Uri $appURL -Method Get -OutFile $appName

$content = [IO.File]::ReadAllText("$PSScriptRoot\latest-version.txt")
Write-Host "  --> Determining latest Git-SCM version: $content "

# Array seperating version details 
$verArr = @($content.split('.'))

# Actually get current Git-SCM Version
$appName="Git-" + $content + "-64-bit.exe"

# Intemediate to calculate the path
# Sample: https://github.com/git-for-windows/git/releases/download/v2.30.0.windows.2/Git-2.30.0.2-64-bit.exe

$calcPath="v" + $verArr[0] + "." + $verArr[1] + "." + $verArr[2] + ".windows." + $verArr[3]
$appURL="https://github.com/git-for-windows/git/releases/download/" + $calcPath + "/" + $appName

Write-Host "  --> Path: $appURL"

Write-Host "  --> Downloading Git $content Installer"

$appCurrent="Git-Current-x64.exe"
Write-Host "  --> Saving as $appCurrent"
Invoke-RestMethod -Uri $appURL -Method Get -OutFile $appCurrent

Write-Host "  --> Git Download Complete."

exit(0)
```

**Comment by dscho** [comment]
Two comments:

- If you use `latest-tag.txt` _in addition_ to `latest-version.txt`, the logic gets _much_ simpler (and more robust).
- You can convert that logic to modify [the script that adjusts `latest-tag.txt` and `latest-version.txt`](https://github.com/git-for-windows/git-for-windows.github.io/blob/8c46971425eca087f740723c100bb0b5f5c0ad73/bump-version.js#L15) to _also_ write, say, `latest-installer.url` and open a PullRequest. That way, you will get something much more reliable, and so will others.

**Comment by dscho** [comment]
> 
> 
> hrm. I seem to be unable to convert this issue to a discussion. 😞

Seems that now I can!

**Reply by hamlibdk** [reply]
Latest PowerShell Logic... 

```
Set-Location -Path $PSScriptRoot

Write-Host ""
Write-Host "* Downloading Git Installer"

# Get current Git-SCM Version
Write-Host ""
$appName="latest-version.txt"
$appURL="https://gitforwindows.org/latest-version.txt"
Invoke-RestMethod -Uri $appURL -Method Get -OutFile $appName
$appName="latest-tag.txt"
$appURL="https://gitforwindows.org/latest-tag.txt"
Invoke-RestMethod -Uri $appURL -Method Get -OutFile $appName

$verContent = [IO.File]::ReadAllText("$PSScriptRoot\latest-version.txt")
Write-Host "  --> Determining latest Git-SCM version: $verContent "

$tagContent = [IO.File]::ReadAllText("$PSScriptRoot\latest-tag.txt")

# Actually get current Git-SCM Version
$appName="Git-" + $verContent + "-64-bit.exe"

# Intermediate to calculate the path
# Sample: https://gitforwindows.org/latest-version.txt  ---> 2.30.1
#         https://gitforwindows.org/latest-tag.txt --------> v2.30.1.windows.1
# Sample: https://github.com/git-for-windows/git/releases/download/v2.30.0.windows.2/Git-2.30.0.2-64-bit.exe
# Sample: https://github.com/git-for-windows/git/releases/download/v2.30.1.windows.1/Git-2.30.1-64-bit.exe

$appURL="https://github.com/git-for-windows/git/releases/download/" + $tagContent + "/" + $appName

Write-Host "  --> Path: $appURL"

Write-Host "  --> Downloading Git $verContent Installer"

$appCurrent="Git-Current-x64.exe"
Write-Host "  --> Saving as $appCurrent"
Invoke-RestMethod -Uri $appURL -Method Get -OutFile $appCurrent

Write-Host "  --> Git Download Complete."

exit(0)
```

**Reply by dscho** [reply]
As I tried to point out gently: a PowerShell script won't cut it.

**Comment by hamlibdk** [comment]
*   As I tried to point out gently: a PowerShell script won't cut it.
Ummm….. Sorry … Pulling data from the two sources of information provided provides the backbone to be able to construct the path (past, present and hopefully future)… It has passed the recent version-bump test. The latest script provided works and has already proven itself. That is not saying process and algorithm will not need a tweak as curve-balls can always appear !

I am no PowerShell advocate or guru having had origins myself back to pre-v1.0.0 Linux kernels … We know the world has resisted PowerShell to the point where Microsoft are taking on Star-Trekkian-Borg like proportions to ensure that “you will assimilate” … On its plus side it is EXTREMELY capable and with its abilities to receive and process data sets as objects …. its capabilities can way outperform Shell scripts !

Sometimes were are left with no fully-capable simple cross-platform, cross-environment alternatives to emulate people and/or environment algorithms …

**Reply by dscho** [reply]
The main problem with that PowerShell script? It is _fragile_, as the filename format/versioning may change at any stage in the future. And since that script is run on each user's computer, it's prone the fall out of date.

Besides, why have a complex script, duplicated many times, when _you can change the website_?

_That_'s why the PowerShell script won't cut it.

**Comment by hamlibdk** [comment]
*   That's why the PowerShell script won't cut it.
No script in any environ will then cut it … if rules and guidance is inconsistent. That has been my point from day 1 ☺ All code needs maintenance.

We need something consistent and perhaps some consistent documentation for versioning … This GIT tool is by far the most advanced tool – with some guidance for those of us who automate pulls.

Enough ! This has gone as far as it can. We will continue to share the code that we use from the JTSDK (and look for us on GitHub) as it helps and guides others.

**Reply by dscho** [reply]
I have mentioned that my preferred way would be to update the script that updates https://gitforwindows.org itself? Maybe that advice wasn't clear enough.

**Comment by hamlibdk** [comment]
No you have not made that clear. Unless you have something constructive to say – rather than toying with people – then please cease further communication.

**Reply by dscho** [reply]
https://github.com/git-for-windows/git/discussions/3020#discussioncomment-341337

Here is the comment.

**Reply by dscho** [reply]
Implement that change, open a PR and I promise I'll work with you to get it merged.

**Reply by dscho** [reply]
Oh, and: could you use threaded replies instead of breaking the threads, please? Thank you.

**Reply by dscho** [reply]
See https://github.com/git-for-windows/git-for-windows.github.io/pull/30.
